File: Graph/Python/graph_am.py
"""
Graph class has an attribute adj_matrix which is a 2D list (or matrix)
representing the edges of the graph.

functions:
- add_edge method adds an edge between two vertices
- remove_edge method removes an edge between two vertices
- is_edge method checks if an edge exists between two vertices
- display method is used to display the adjacency matrix.
"""
class Graph:
    def __init__(self, num_vertices):
        self.num_vertices = num_vertices
        self.adj_matrix = [[0 for _ in range(num_vertices)] for _ in range(num_vertices)]

    def add_edge(self, v1, v2):
        self.adj_matrix[v1][v2] = 1
        self.adj_matrix[v2][v1] = 1

    def remove_edge(self, v1, v2):
        if self.is_edge(v1, v2):
            self.adj_matrix[v1][v2] = 0
            self.adj_matrix[v2][v1] = 0

    def is_edge(self, v1, v2):
        return True if self.adj_matrix[v1][v2] > 0 else False

    def display(self):
        for row in self.adj_matrix:
            for val in row:
                print(val, end=" ")
            print()

# Example usage:
g = Graph(5)
g.add_edge(0, 1)
g.add_edge(0, 2)
g.add_edge(1, 2)
g.add_edge(2, 0)
g.add_edge(2, 3)

print("Adjacency Matrix:")
g.display()

# Output:
# Adjacency Matrix:
# 0 1 1
# 1 0 1
# 1 1 0
# 0 0 0
# 0 0 0

File: Graph/C/graph_am.c
/*
Adjacency Matrix

Simple undirected graph with 4 vertices and a few edges between them

output:
0 1 1 0
1 0 1 0
1 1 0 1
0 0 1 0
*/
#include <stdio.h>

#define V 4 // Number of vertices

// Function to create a new graph
int **createGraph() {
    int **graph = (int **)malloc(V * sizeof(int *));
    for (int i = 0; i < V; i++) {
        graph[i] = (int *)malloc(V * sizeof(int));
        for (int j = 0; j < V; j++) {
            graph[i][j] = 0;
        }
    }
    return graph;
}

// Function to add an edge to the graph
void addEdge(int **graph, int src, int dest) {
    graph[src][dest] = 1;
    graph[dest][src] = 1; // For an undirected graph
}

// Function to print the graph
void printGraph(int **graph) {
    for (int i = 0; i < V; i++) {
        for (int j = 0; j < V; j++) {
            printf("%d ", graph[i][j]);
        }
        printf("\n");
    }
}

int main() {
    int **graph = createGraph();

    // Add edges to the graph
    addEdge(graph, 0, 1);
    addEdge(graph, 0, 2);
    addEdge(graph, 1, 2);
    addEdge(graph, 2, 3);

    // Print the graph
    printGraph(graph);

    return 0;
}
